### DISCUSSION TITLE
Migration to a different host

### DISCUSSION THREAD

**Discussion by Err2101** [ubergeek77/Lemmy-Easy-Deploy#83] [Q&A] [answered]
Hi all, I imagine this is not the correct place to post this but I'm not sure where to ask for help as it seems to be a specific question about this project. 

I want to move from a current Ubuntu host to Debian ( i am moving all my hosts to debian to standardize) 

So on the OG lemmy docker host I stopped all containers, ran docker commit for all off them, then docker save to export the images to .tar files. I used rsync to move the files at /var/lib/docker/volumes/* to the new host. I then moved over all the .tar files to the new host and used docker load to load the images, then docker run -d <imageid> to start the containers.

It looks like they all came up besides ghcr.io/ubergeek77/lemmy:0.18.3. I tried to just do docker run <imageid> to try to get an error that might lead me somewhere, but just get :

Error response from daemon: pull access denied for 4a8435c0512a, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.

So Im just looking for some help from someone who might have moved docker containers between hosts before, or anyone who's done what I'm doing, what did you do? 

Thank you in advance

**Comment by ubergeek77** [comment]
I don't know why you can't pull that image, but it is 5 months old at this point. The latest version is 0.19. Please try running that.

On the new machine, if you run `docker volume ls`, do you see all of the Lemmy Easy Deploy volumes? If so, all you need to do is run `deploy.sh` and it will get the images it needs. Just make sure your Lemmy Easy Deploy folder was copied over too.

If you are running a very old version of Lemmy Easy Deploy from before I added the update detector, you may need to update the script as well.

**Comment by Err2101** [comment]
Yes I see those volumes. I must have done something wrong - I ran the deploy.sh script and got your big warning box about my deployment not being tracked, but data volumes already exist. I did notice that a container ls does not have the same image names as the old host, just random codes -

`CONTAINER ID   IMAGE          COMMAND                  CREATED          STATUS          PORTS                NAMES
d8e255fc4575   a1564b0b8454   "docker-entrypoint.s…"   26 minutes ago   Up 26 minutes   5432/tcp             clever_ptolemy
7b957ccaa985   1ec959b37b66   "/sbin/tini -- /usr/…"   27 minutes ago   Up 27 minutes   6669/tcp, 8080/tcp   reverent_kowalevski
d7352e203cce   797809870dbf   "docker-entrypoint.s…"   27 minutes ago   Up 27 minutes   1234/tcp             confident_austin
5d9594d3e759   a1564b0b8454   "docker-entrypoint.s…"   41 minutes ago   Up 41 minutes   5432/tcp             interesting_borg
`

**Reply by Err2101** [reply]
Sorry that looks terrible, here's an image.
![image](https://github.com/ubergeek77/Lemmy-Easy-Deploy/assets/132849276/4ea16f96-29a2-4ab7-9ebd-f45deb71b0cd)

**Comment by ubergeek77** [comment, marked as answer]
You are seeing that warning because you did not copy the entire Lemmy Easy Deploy folder from your previous host. That folder has credentials and other files in it that your deployment needs.

I also think that trying to copy the containers was not the best idea, there is no telling what could go wrong.

My recommendation is to start over:

- On the new machine, delete all docker volumes, images, and containers related to Lemmy Easy Deploy.
- Re-copy ***only*** volumes from the previous host, not images and not containers
- Do not attempt to start any containers manually
- Copy your old `Lemmy-Easy-Deploy` folder from your old host. I recommend removing the folder from this new host which failed, then getting a fresh copy of the folder.
- Ensure the `live` folder already exists inside the `Lemmy-Easy-Deploy` folder before doing anything else
- Ensure Lemmy-Easy-Deploy is up to date (1.3.2) before continuing
  - Don't just download an updated `deploy.sh`, make sure every file in this repository is up to date. If you still have your .git folder, you may be able to just run `git pull `

Once you have done all of that, try running ./deploy.sh again.

**Reply by Err2101** [reply]
Okay I'll do all of this right now. I really appreciate your help. Docker is new to me I have previously just been using LXCs for everything :)

**Reply by Err2101** [reply]
This seemed to work but I did get these warnings:

                                                                                                                                                                                                        docker:default
Warning: No resource found to remove for project "lemmy-easy-deploy".
WARN[0000] volume "lemmy-easy-deploy_pictrs_data" already exists but was not created by Docker Compose. Use `external: true` to use an existing volume 
WARN[0000] volume "lemmy-easy-deploy_postgres_data" already exists but was not created by Docker Compose. Use `external: true` to use an existing volume 
WARN[0000] volume "lemmy-easy-deploy_caddy_data" already exists but was not created by Docker Compose. Use `external: true` to use an existing volume 
WARN[0000] volume "lemmy-easy-deploy_caddy_config" already exists but was not created by Docker Compose. Use `external: true` to use an existing volume 

Where would I set this option?

**Reply by ubergeek77** [reply]
Interesting. This is because the volumes don't have tags to tell Compose that it was created by Compose. I would have thought that would be copied over too.

You can safely ignore this warning, your deployment will continue to work fine. That instruction to set `external` to `true` isn't really relevant to this situation.

You will continue to see this warning on upgrades, but Compose will still use the correct volumes.

**Reply by Err2101** [reply]
Okay looks like it all started, but I'm getting this in browser. I imagine I'll have to go look at lemmy docs to figure this one out:

![image](https://github.com/ubergeek77/Lemmy-Easy-Deploy/assets/132849276/29634ec2-875f-4a02-a8c4-455444f05a7b)


Thank you again for all your help so far!!

**Reply by ubergeek77** [reply]
You should have gotten a little red popover on the bottom left saying "Error: Lemmy is upgrading"

You will see this while Lemmy performs database migrations. Don't stop the containers, don't restart them, just give it about 10-15 minutes and it will start working again when it's done :)

**Reply by Err2101** [reply]
Great I will give it some time. Thanks!

**Reply by ubergeek77** [reply]
You're very welcome! Feel free to respond if you have any more issues!

**Reply by Err2101** [reply]
It's working! Thank you! 

Should we put your answer as a wiki post, or something?

**Reply by ubergeek77** [reply]
I think for now, I am comfortable with pointing people here.

I'm not maintaining a wiki for this repository at the moment (and GitHub doesn't support PRs for wikis, which is pretty lame). However, if you were interested in taking my instructions and writing a more fleshed out guide around it based on your experience doing this, you're more than welcome to write it, and I could put it in a wiki page and credit you as the author.

**Reply by Err2101** [reply]
What are PRs for wikis? 

Here is what I have so far. I'm not sure what some good guidelines for writing articles like this are, since install directories and such are so non-standard. Let me know if I can help any other way, I appreciate you!

This guide means to outline the basic process of migrating a Lemmy-Easy-Deploy deployment to a different host. 

Before following this guide, be sure the following is ready:
Docker installed on the new host
	https://docs.docker.com/engine/install/#server
A way to move files between the two servers. The below contains a way to do so via rsync over LAN, but WinSCP or some other method can be used if there isn't a direct connection between the two.


On the old host:

Moving Volumes:
Stop the lemmy easy deploy containers using deploy.sh in the Lemmy-Easy-Deploy directory:
./deploy.sh -s
It is recommend to use rsync when moving this data in case something happens during the copy and the process is restarted. Consider compressing this data if there is an excess amount:
rsync -arzvh /var/lib/docker/volumes/lemmy* user@newhostip:/var/lib/docker/volumes/

Moving Lemmy-Easy-Deploy data:

Before moving, update Lemmy-Easy-Deploy to the newest version:
cd into your Lemmy-Easy-Deploy directory, and use git-pull:
git pull Lemmy-Easy-Deploy

Copy the entire Lemmy-Easy-Deploy directory via rsync or scp to the new host:
rsync -arzvh Lemmy-Easy-Deploy/* user@newhostip:/path/Lemmy-Easy-Deploy/

On the new host:

Ensure that the "live" directory in the Lemmy-Easy-Deploy directory is moved over from the old server properly and contains the same files. 

Run ./deploy.sh in the Lemmy-Easy-Deploy directory. 
The script will begin to deploy the docker containers for Lemmy. 
The following warnings may appear. They are caused by copying the volume data to the new host manually and not creating them through docker compose. They can be ignored:
Warning: No resource found to remove for project "lemmy-easy-deploy".
WARN[0000] volume "lemmy-easy-deploy_pictrs_data" already exists but was not created by Docker Compose. Use external: true to use an existing volume
WARN[0000] volume "lemmy-easy-deploy_postgres_data" already exists but was not created by Docker Compose. Use external: true to use an existing volume
WARN[0000] volume "lemmy-easy-deploy_caddy_data" already exists but was not created by Docker Compose. Use external: true to use an existing volume
WARN[0000] volume "lemmy-easy-deploy_caddy_config" already exists but was not created by Docker Compose. Use external: true to use an existing volume

Once the startup is complete, the following may also appear on the main webpage. This is normal as Lemmy is upgrading. Wait 10-15 minutes(Do not restart or shut down any containers!) and the upgrade will complete and the normal Lemmy webpage will be displayed! 
![image](https://github.com/ubergeek77/Lemmy-Easy-Deploy/assets/132849276/e633f5d6-c6ce-4b86-8e05-6d89866c0dba)


This should be everything needed to migrate Lemmy-Easy-Deploy to a new host!

**Reply by ubergeek77** [reply]
I've copied your guide and incorporated it into a new Wiki page, thank you so much for writing it!

https://github.com/ubergeek77/Lemmy-Easy-Deploy/wiki/Migrating-a-Lemmy-Easy-Deploy-deployment-from-one-machine-to-another

> What are PRs for wikis?

Ah, I just meant, it's not possible to submit a Pull Request on GitHub for Wiki pages. I.e., there's no formal way for you to write a wiki page first, then ask me to add it to the wiki. The only option for adding to Wikis would be to coordinate like we just did :p

**Reply by Err2101** [reply]
Huh. That's surprising. Well I appreciate your help! and happy I could write something up to hopefully help others in the future!
